src/main/java/org/opensrp/repository/postgres/PlanRepositoryImpl.java

```diff
@@ -0,0 +1,213 @@
+
+        String id = retrievePrimaryKey(plan);
+        if (id == null) {
+            return;
+        }
+
+        Plan pgPlan = convert(plan);
+        pgPlan.setIsDeleted(true);
+        planMapper.updateByPrimaryKey(pgPlan);
+    }
+
+    public List<PlanDefinition> getPlansByServerVersionAndOperationalArea(Long serverVersion, String operationalAreaId) {
+        PlanExample planExample = new PlanExample();
+        planExample.createCriteria().andServerVersionGreaterThanOrEqualTo(serverVersion);
+        List<Plan> plans = planMapper.selectMany(planExample, operationalAreaId, 0, DEFAULT_FETCH_SIZE);
+
+        return convert(plans);
+    }
+
+    @Override
+    protected String retrievePrimaryKey(PlanDefinition plan) {
+        Object uniqueId = getUniqueField(plan);
+        if (uniqueId == null) {
+            return null;
+        }
+
+        String identifier = uniqueId.toString();
+        PlanDefinition pgPlan = get(identifier);
+
+        return pgPlan == null ? null : pgPlan.getIdentifier();
+    }
+
+    @Override
+    protected Object getUniqueField(PlanDefinition plan) {
+        return plan == null ? null : plan.getIdentifier();
+    }
+
+    private PlanDefinition convert(Plan pgPlan) {
+        if (pgPlan == null || pgPlan.getJson() == null || !(pgPlan.getJson() instanceof PlanDefinition)) {
+            return null;
+        }
+        return (PlanDefinition) pgPlan.getJson();
+    }
+
+    private Plan convert(PlanDefinition plan) {
+        if (plan == null) {
+            return null;
+        }
+        Plan pgPlan = new Plan();
+        pgPlan.setId(plan.getIdentifier());
+        pgPlan.setJson(plan);
+        pgPlan.setServerVersion(plan.getServerVersion());
+
+        return pgPlan;
+    }
+
+    private List<PlanDefinition> convert(List<Plan> pgPlans) {
+        List<PlanDefinition> plans = new ArrayList<>();
+        if (isEmptyList(pgPlans)) {
+            return plans;
+        }
+        for(Plan pgPlan : pgPlans) {
+            plans.add(convert(pgPlan));
+        }
+        return plans;
+    }
+
+    private void insertPlanMetadata(PlanDefinition plan) {
+        if (isEmptyList(plan.getJurisdiction())) {
+            return;
+        }
+        for (Jurisdiction jurisdiction : plan.getJurisdiction()) {
+           insert(jurisdiction, plan);
+        }
+    }
+
+    private void insert(Jurisdiction jurisdiction, PlanDefinition plan) {
+        PlanMetadata planMetadata = new PlanMetadata();
+        planMetadata.setOperationalAreaId(jurisdiction.getCode());
+        planMetadata.setPlanId(plan.getIdentifier());
+        planMetadata.setIsDeleted(false);
+        planMetadataMapper.insert(planMetadata);
+    }
+
+    private void updatePlanMetadata(PlanDefinition plan) {
+        Set<String> operationalAreas = new HashSet<>();
+        for (Jurisdiction jurisdiction : plan.getJurisdiction()) {
+            if (retrievePrimaryKey(plan) == null) {
+                insert(jurisdiction, plan);
+            }
```

Comment:
Add an else and check if the operational area has been removed from the plan, in that case delete the PlanMetadata